# IC 4.5. Data Consistency Checks

### 0. List of Contents
#### 1. Importing Libraries and Dataframes
#### 2 Handling Mixed-Type Data
#### 3. Handling Missing Values
#### 4. Checking for Duplicates
#### 5. Consistency Checks on Prod DF
#### 6. Consistency Checks on Ords DF
#### 7. Exporting DF

### 1. Importing Libraries and Dataframes

In [2]:
#importing libraries
import pandas as pd
import numpy as np
import os

In [3]:
#creating path
path = r'C:\Users\User\Documents\Tanulok\CareerFoundry\2. Data Immersion\4. Python\Instacart Basket Analysis\02 Data'

In [4]:
#importing orders
df_ords = pd.read_csv(os.path.join(path, 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [5]:
#importing products
df_prods = pd.read_csv(os.path.join(path, 'Original Data', 'products.csv'), index_col = False)

### 2 Handling Mixed-Type Data

In [6]:
#creating a dataframe
df_test = pd.DataFrame()

In [7]:
# creating a mixed type column
df_test['mix']= ['a', 'b', 1, True]

In [8]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [9]:
#checking for mixed types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [10]:
#chaging data type
df_test['mix'] = df_test['mix'].astype('str')

### 3. Handling Missing Values

In [11]:
#finding missing values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [12]:
#creating subset with missing values
df_nan=df_prods[df_prods['product_name'].isnull() == True]

In [13]:
#checking df_nan
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [14]:
#counting rows to see if missing values can be removed
df_prods.shape

(49693, 5)

In [15]:
#creating new data set without missing values
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [16]:
#counting rows for new prod df
df_prods_clean.shape

(49677, 5)

### 4. Checking for Duplicates

In [17]:



#finding duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [18]:
#pulling up duplicates
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [19]:
#checking number of rows before removing duplicates
df_prods_clean.shape

(49677, 5)

In [20]:
#creating new data subset without duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [21]:
#checking df without duplicates
df_prods_clean_no_dups.shape

(49672, 5)

In [22]:
#exporting clean data set
df_prods_clean_no_dups.to_csv(os.path.join(path,'Prepared Data', 'products_checked.csv'))

### 5. Consistency Checks on Prod DF

In [23]:
#Running describe
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

#### Interpretation of the results
- product_id starts (min) at 1 and goes until (max) 49688, but we have 49693 (count) values -> could this mean duplicates
- min, max and median/50% looks correct for aisle_id and department_id
- for price the maximum 99 999 seems too high compared to the other values

In [24]:
#Checking for mixed-type data
for col in df_prods.columns.tolist():
  weird = (df_prods[[col]].applymap(type) != df_prods[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_prods[weird]) > 0:
    print (col)

product_name


In [26]:
#As product_name is mixed-type, it needs to be changed to STR, because names are always made up of letters
df_prods['product_name'] = df_prods['product_name'].astype('str')

### 6. Consistency Checks on Ords DF

In [28]:
#Correction for 3. mixed-type data to be checked for df_ords
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

#### df_ords doesn't include mixed-type data

In [29]:
# Run a check for missing values in your df_ords dataframe
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
eval_set                       0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

#### days_since_prior_order has high nan values, probably for the first order, when there isn't a previous order


In [33]:
#Checking number of rows to see if my assumption about the missing values can be true
df_ords.shape

(3421083, 8)

In [34]:
#Checking df_ords with describe
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

#### Max user_id = Missing values in days_since_prior_order, hence missing values will remain

In [36]:
#Run a check for duplicate values in your df_ords data. 
df_dups_ords = df_ords[df_ords.duplicated()]

In [38]:
#Checking this new subset of ords dups
df_dups_ords

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

#### There are no duplicates, hence they don't need to be addressed

### 7. Exporting DF

In [39]:
# Exporting clean df_ords
df_ords.to_csv(os.path.join(path,'Prepared Data', 'orders_checked.csv'))